Gradle: `:quarkusDependenciesBuild` uses this output of `:jandex` without declaring dependency

### Describe the bug

A new minimal Quarkus project using Gradle 8.10. as build system fails to build due to a missing dependency declaration to the `org.kordamp.gradle.jandex` plugin.

### Expected behavior

Build should be successful.

When removing the Jandex plugin from `build.gradle.kts`, the build succeeds:
```
❯ ./gradlew build

BUILD SUCCESSFUL in 1s
7 actionable tasks: 6 executed, 1 up-to-date
```

### Actual behavior

```
❯ ./gradlew build
> Task :quarkusDependenciesBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':quarkusDependenciesBuild' (type 'QuarkusBuildDependencies').
  - Gradle detected a problem with the following location: '/path/to/quarkus-jandex/build/resources/main'.
    
    Reason: Task ':quarkusDependenciesBuild' uses this output of task ':jandex' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':jandex' as an input of ':quarkusDependenciesBuild'.
      2. Declare an explicit dependency on ':jandex' from ':quarkusDependenciesBuild' using Task#dependsOn.
      3. Declare an explicit dependency on ':jandex' from ':quarkusDependenciesBuild' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.10.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 1s
7 actionable tasks: 5 executed, 2 up-to-date
```

### How to Reproduce?

`build.gradle.kts`:
```kotlin
plugins {
	java
	id("org.kordamp.gradle.jandex") version "2.0.0"
	id("io.quarkus") version "3.15.1"
}

repositories {
	mavenCentral()
}

dependencies {
	implementation(enforcedPlatform("io.quarkus.platform:quarkus-bom:3.15.1"))
}

group = "com.example"
version = "1.0.0-SNAPSHOT"
```

`settings.gradle.kts`:
```kotlin
pluginManagement {
	repositories {
		mavenCentral()
		gradlePluginPortal()
	}
}
rootProject.name = "jandex-dependency"
```

There are no other source files in `src/` whatsoever.

### Output of `uname -a` or `ver`

Darwin macbook.local 23.6.0 Darwin Kernel Version 23.6.0: Wed Jul 31 20:48:04 PDT 2024; root:xnu-10063.141.1.700.5~1/RELEASE_ARM64_T6030 arm64

### Output of `java -version`

openjdk 21.0.5 2024-10-15 LTS

### Quarkus version or git rev

3.15.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.10.2

### Additional information

_No response_

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
@@ -249,9 +249,9 @@ public class QuarkusPlugin implements Plugin<Project> {
                 QuarkusBuildDependencies.class,
                 task -> {
                     configureQuarkusBuildTask(project, task, quarkusBuildAppModelTask, serviceProvider);
-
                     task.getOutputs().doNotCacheIf("Dependencies are never cached", t -> true);
                 });
+        project.afterEvaluate(evaluated -> addDependencyOnJandexIfConfigured(evaluated, quarkusBuildDependencies));
 
         Property<Boolean> cacheLargeArtifacts = quarkusExt.getCacheLargeArtifacts();
 
@@ -646,7 +646,7 @@ public class QuarkusPlugin implements Plugin<Project> {
     }
 
     private void setupQuarkusBuildTaskDeps(Project project, Project dep, Set<String> visited) {
-        if (!visited.add(dep.getGroup() + ":" + dep.getName())) {
+        if (!visited.add(dep.getPath())) {
             return;
         }
 
@@ -663,16 +663,22 @@ public class QuarkusPlugin implements Plugin<Project> {
                 });
 
         getLazyTask(project, QUARKUS_DEV_TASK_NAME).ifPresent(quarkusDev -> {
-            for (String taskName : new String[] { JavaPlugin.PROCESS_RESOURCES_TASK_NAME,
-                    // This is the task of the 'org.kordamp.gradle.jandex' Gradle plugin
-                    "jandex",
-                    // This is the task of the 'com.github.vlsi.jandex' Gradle plugin
-                    "processJandexIndex" }) {
-                getLazyTask(dep, taskName).ifPresent(t -> quarkusDev.configure(qd -> qd.dependsOn(t)));
-            }
+            getLazyTask(project, JavaPlugin.PROCESS_RESOURCES_TASK_NAME)
+                    .ifPresent(t -> quarkusDev.configure(qd -> qd.dependsOn(t)));
+            addDependencyOnJandexIfConfigured(dep, quarkusDev);
         });
 
         visitProjectDependencies(project, dep, visited);
+    }
+
+    private void addDependencyOnJandexIfConfigured(Project project, TaskProvider<? extends Task> quarkusTask) {
+        for (String taskName : new String[] {
+                // This is the task of the 'org.kordamp.gradle.jandex' Gradle plugin
+                "jandex",
+                // This is the task of the 'com.github.vlsi.jandex' Gradle plugin
+                "processJandexIndex" }) {
+            getLazyTask(project, taskName).ifPresent(t -> quarkusTask.configure(qd -> qd.mustRunAfter(t)));
+        }
     }
 
     protected void visitProjectDependencies(Project project, Project dep, Set<String> visited) {
@@ -689,10 +695,10 @@ public class QuarkusPlugin implements Plugin<Project> {
                     .forEach(d -> {
                         Project depProject = null;
 
-                        if (d instanceof ProjectDependency) {
-                            depProject = dep.project(((ProjectDependency) d).getPath());
-                        } else if (d instanceof ExternalModuleDependency) {
-                            depProject = ToolingUtils.findIncludedProject(project, (ExternalModuleDependency) d);
+                        if (d instanceof ProjectDependency projectDep) {
+                            depProject = dep.project(projectDep.getPath());
+                        } else if (d instanceof ExternalModuleDependency externalModuleDep) {
+                            depProject = ToolingUtils.findIncludedProject(project, externalModuleDep);
                         }
 
                         if (depProject == null) {
@@ -711,13 +717,9 @@ public class QuarkusPlugin implements Plugin<Project> {
     private void visitLocalProject(Project project, Project localProject, Set<String> visited) {
         // local dependency, so we collect also its dependencies
         visitProjectDep(project, localProject, visited);
-
-        ExtensionDependency<?> extensionDependency = DependencyUtils
-                .getExtensionInfoOrNull(project, localProject);
-
-        if (extensionDependency instanceof ProjectExtensionDependency) {
-            visitProjectDep(project,
-                    ((ProjectExtensionDependency) extensionDependency).getDeploymentModule(), visited);
+        ExtensionDependency<?> extensionDependency = DependencyUtils.getExtensionInfoOrNull(project, localProject);
+        if (extensionDependency instanceof ProjectExtensionDependency projectExtDep) {
+            visitProjectDep(project, projectExtDep.getDeploymentModule(), visited);
         }
     }
 

--- a/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/application/build.gradle
+++ b/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/application/build.gradle
@@ -1,5 +1,6 @@
 plugins {
     id 'io.quarkus'
+    id 'org.kordamp.gradle.jandex' // this plugin is applied to make the dependencies on the Jandex tasks are properly setup and the build doesn't fail
 }
 
 dependencies {

--- a/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/build.gradle
+++ b/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/build.gradle
@@ -39,7 +39,7 @@ subprojects {
     }
 
     dependencies {
-        implementation 'io.quarkus:quarkus-resteasy'
+        implementation 'io.quarkus:quarkus-rest'
 
         implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
     }

--- a/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/common/build.gradle
+++ b/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/common/build.gradle
@@ -1,7 +1,7 @@
 plugins {
     id 'io.quarkus'
     id 'java-library'
-    id "org.kordamp.gradle.jandex" version "1.1.0"
+    id 'org.kordamp.gradle.jandex'
 }
 
 dependencies {

--- a/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/gradle.properties
@@ -1,2 +1,3 @@
 quarkusPlatformArtifactId=quarkus-bom
 quarkusPlatformGroupId=io.quarkus
+kordampJandexVersion=2.1.0

--- a/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/settings.gradle
+++ b/integration-tests/gradle/src/main/resources/jandex-basic-multi-module-project-kordamp/settings.gradle
@@ -12,6 +12,7 @@ pluginManagement {
 
     plugins {
         id 'io.quarkus' version "${quarkusPluginVersion}"
+        id 'org.kordamp.gradle.jandex' version "${kordampJandexVersion}"
     }
 }
 

